File: WFM-Planning-Suite/web/src/core/erlang/engine.ts
import type { ErlangModel } from '../models'

export interface ErlangKPIs {
  agents: number
  trafficErlangs: number
  slaPct: number
  asaSeconds: number
  abandonmentPct: number
  occupancyPct: number
  model: ErlangModel
}

export function erlangB(agents: number, traffic: number): number {
  if (agents <= 0) return 1
  if (traffic <= 0) return 0
  let invB = 1
  for (let k = 1; k <= agents; k++) invB = 1 + (k / traffic) * invB
  return 1 / invB
}

export function erlangC(agents: number, traffic: number): number {
  if (agents <= 0) return 1
  if (traffic <= 0) return 0
  if (agents <= traffic) return 1
  const b = erlangB(agents, traffic)
  const denom = agents - traffic * (1 - b)
  if (denom <= 0) return 1
  return (agents * b) / denom
}

export function slaErlangC(agents: number, traffic: number, aht: number, slaTime: number): number {
  if (traffic <= 0 || agents <= 0) return 100
  if (agents <= traffic) return 0
  const c = erlangC(agents, traffic)
  if (aht <= 0) return 0
  const exp = -(agents - traffic) * (slaTime / aht)
  return Math.max(0, Math.min(100, (1 - c * Math.exp(exp)) * 100))
}

export function asaErlangC(agents: number, traffic: number, aht: number): number {
  if (traffic <= 0 || agents <= 0) return 0
  if (agents <= traffic) return Infinity
  const c = erlangC(agents, traffic)
  return (c * aht) / (agents - traffic)
}

export function abandonmentErlangA(agents: number, traffic: number, aht: number, patience: number): number {
  if (traffic <= 0 || patience <= 0) return 0
  if (agents <= traffic) return 100
  const c = erlangC(agents, traffic)
  if (aht <= 0) return 0
  const factor = (agents - traffic) / aht
  if (factor <= 0) return 0
  return Math.max(0, Math.min(100, c * (factor / (factor + 1 / patience)) * 100))
}

export function slaErlangA(agents: number, traffic: number, aht: number, slaTime: number, patience: number): number {
  const abandon = abandonmentErlangA(agents, traffic, aht, patience) / 100
  const base = slaErlangC(agents, traffic, aht, slaTime) / 100
  return Math.max(0, Math.min(100, base * (1 - abandon * 0.5) * 100))
}

function occupancy(traffic: number, agents: number): number {
  if (agents <= 0) return 100
  return Math.max(0, Math.min(100, (traffic / agents) * 100))
}

export function kpisForAgents(
  agents: number, volume: number, intervalSec: number, aht: number,
  model: ErlangModel, slaTime = 20, patience = 120,
): ErlangKPIs {
  const traffic = volume > 0 ? (volume * aht) / intervalSec : 0
  const n = Math.max(1, agents)

  if (model === 'erlang_b') {
    const block = erlangB(n, traffic) * 100
    return { agents: n, trafficErlangs: traffic, slaPct: Math.max(0, 100 - block), asaSeconds: 0, abandonmentPct: block, occupancyPct: occupancy(traffic, n), model }
  }
  if (model === 'erlang_c') {
    return { agents: n, trafficErlangs: traffic, slaPct: slaErlangC(n, traffic, aht, slaTime), asaSeconds: asaErlangC(n, traffic, aht), abandonmentPct: 0, occupancyPct: occupancy(traffic, n), model }
  }
  return { agents: n, trafficErlangs: traffic, slaPct: slaErlangA(n, traffic, aht, slaTime, patience), asaSeconds: asaErlangC(n, traffic, aht), abandonmentPct: abandonmentErlangA(n, traffic, aht, patience), occupancyPct: occupancy(traffic, n), model }
}

export function requiredAgents(
  volume: number, intervalSec: number, aht: number, model: ErlangModel,
  slaTarget = 80, slaTime = 20, patience = 120, maxAgents = 500,
): [number, ErlangKPIs] {
  if (volume <= 0) return [0, kpisForAgents(0, 0, intervalSec, aht, model, slaTime, patience)]
  const traffic = (volume * aht) / intervalSec
  const start = Math.max(1, Math.ceil(traffic))
  for (let n = start; n <= maxAgents; n++) {
    const kpi = kpisForAgents(n, volume, intervalSec, aht, model, slaTime, patience)
    if (model === 'erlang_b') {
      if (kpi.abandonmentPct <= 100 - slaTarget) return [n, kpi]
    } else if (kpi.slaPct >= slaTarget) return [n, kpi]
  }
  return [maxAgents, kpisForAgents(maxAgents, volume, intervalSec, aht, model, slaTime, patience)]
}

File: WFM-Planning-Suite/src/core/erlang/engine.py
"""Single Erlang A/B/C engine — one implementation per model."""

from __future__ import annotations

import math
from dataclasses import dataclass

from core.models import ErlangModel


@dataclass(frozen=True)
class ErlangKPIs:
    agents: int
    traffic_erlangs: float
    sla_pct: float
    asa_seconds: float
    abandonment_pct: float
    occupancy_pct: float
    model: ErlangModel


def erlang_b(agents: int, traffic: float) -> float:
    """Blocking probability Erlang B."""
    if agents <= 0:
        return 1.0
    if traffic <= 0:
        return 0.0
    inv_b = 1.0
    for k in range(1, agents + 1):
        inv_b = 1.0 + (k / traffic) * inv_b
    return 1.0 / inv_b


def erlang_c(agents: int, traffic: float) -> float:
    """Probability of wait Erlang C."""
    if agents <= 0:
        return 1.0
    if traffic <= 0:
        return 0.0
    if agents <= traffic:
        return 1.0
    b = erlang_b(agents, traffic)
    denom = agents - traffic * (1.0 - b)
    if denom <= 0:
        return 1.0
    return (agents * b) / denom


def sla_erlang_c(
    agents: int,
    traffic: float,
    aht_seconds: float,
    sla_time_seconds: float,
) -> float:
    """Service level for Erlang C: % answered within SLA time."""
    if traffic <= 0 or agents <= 0:
        return 100.0
    if agents <= traffic:
        return 0.0
    c = erlang_c(agents, traffic)
    if aht_seconds <= 0:
        return 0.0
    exponent = -(agents - traffic) * (sla_time_seconds / aht_seconds)
    return max(0.0, min(100.0, (1.0 - c * math.exp(exponent)) * 100.0))


def asa_erlang_c(agents: int, traffic: float, aht_seconds: float) -> float:
    """Average speed of answer for Erlang C."""
    if traffic <= 0 or agents <= 0:
        return 0.0
    if agents <= traffic:
        return float("inf")
    c = erlang_c(agents, traffic)
    return (c * aht_seconds) / (agents - traffic)


def abandonment_erlang_a(
    agents: int,
    traffic: float,
    aht_seconds: float,
    patience_seconds: float,
) -> float:
    """Abandonment rate for Erlang A (exponential patience)."""
    if traffic <= 0 or patience_seconds <= 0:
        return 0.0
    if agents <= traffic:
        return 100.0
    c = erlang_c(agents, traffic)
    if aht_seconds <= 0:
        return 0.0
    factor = (agents - traffic) / aht_seconds
    if factor <= 0:
        return 0.0
    return max(0.0, min(100.0, c * (factor / (factor + 1.0 / patience_seconds)) * 100.0))


def sla_erlang_a(
    agents: int,
    traffic: float,
    aht_seconds: float,
    sla_time_seconds: float,
    patience_seconds: float,
) -> float:
    """Service level for Erlang A with abandonment."""
    if traffic <= 0:
        return 100.0
    abandon = abandonment_erlang_a(agents, traffic, aht_seconds, patience_seconds) / 100.0
    base_sla = sla_erlang_c(agents, traffic, aht_seconds, sla_time_seconds) / 100.0
    return max(0.0, min(100.0, base_sla * (1.0 - abandon * 0.5) * 100.0))


def occupancy(traffic: float, agents: int) -> float:
    if agents <= 0:
        return 100.0
    return max(0.0, min(100.0, (traffic / agents) * 100.0))


def kpis_for_agents(
    agents: int,
    volume: float,
    interval_seconds: float,
    aht_seconds: float,
    model: ErlangModel,
    sla_time_seconds: float = 20.0,
    patience_seconds: float = 120.0,
) -> ErlangKPIs:
    """Compute KPIs for a fixed agent count using the selected model."""
    if interval_seconds <= 0:
        interval_seconds = 1800.0
    traffic = (volume * aht_seconds) / interval_seconds if volume > 0 else 0.0
    agents = max(1, agents)

    if model == ErlangModel.B:
        block = erlang_b(agents, traffic) * 100.0
        return ErlangKPIs(
            agents=agents,
            traffic_erlangs=traffic,
            sla_pct=max(0.0, 100.0 - block),
            asa_seconds=0.0,
            abandonment_pct=block,
            occupancy_pct=occupancy(traffic, agents),
            model=model,
        )

    if model == ErlangModel.C:
        return ErlangKPIs(
            agents=agents,
            traffic_erlangs=traffic,
            sla_pct=sla_erlang_c(agents, traffic, aht_seconds, sla_time_seconds),
            asa_seconds=asa_erlang_c(agents, traffic, aht_seconds),
            abandonment_pct=0.0,
            occupancy_pct=occupancy(traffic, agents),
            model=model,
        )

    # Erlang A
    return ErlangKPIs(
        agents=agents,
        traffic_erlangs=traffic,
        sla_pct=sla_erlang_a(agents, traffic, aht_seconds, sla_time_seconds, patience_seconds),
        asa_seconds=asa_erlang_c(agents, traffic, aht_seconds),
        abandonment_pct=abandonment_erlang_a(agents, traffic, aht_seconds, patience_seconds),
        occupancy_pct=occupancy(traffic, agents),
        model=model,
    )


def required_agents(
    volume: float,
    interval_seconds: float,
    aht_seconds: float,
    model: ErlangModel,
    sla_target_pct: float = 80.0,
    sla_time_seconds: float = 20.0,
    patience_seconds: float = 120.0,
    max_agents: int = 500,
) -> tuple[int, ErlangKPIs]:
    """Find minimum agents to meet SLA target."""
    if volume <= 0:
        kpi = kpis_for_agents(0, 0, interval_seconds, aht_seconds, model, sla_time_seconds, patience_seconds)
        return 0, kpi

    traffic = (volume * aht_seconds) / interval_seconds
    start = max(1, int(math.ceil(traffic)))

    for n in range(start, max_agents + 1):
        kpi = kpis_for_agents(n, volume, interval_seconds, aht_seconds, model, sla_time_seconds, patience_seconds)
        if model == ErlangModel.B:
            if kpi.abandonment_pct <= (100.0 - sla_target_pct):
                return n, kpi
        elif kpi.sla_pct >= sla_target_pct:
            return n, kpi

    kpi = kpis_for_agents(max_agents, volume, interval_seconds, aht_seconds, model, sla_time_seconds, patience_seconds)
    return max_agents, kpi
